pkg/trafficrouting/manager.go

```diff
@@ -217,9 +223,16 @@ func (m *Manager) FinalisingTrafficRouting(c *TrafficRoutingContext, onlyRestore
 	}
 
 	cService := &corev1.Service{ObjectMeta: metav1.ObjectMeta{Namespace: c.Namespace, Name: cServiceName}}
+	// In rollout failure case, no canary-service will be created, but stable service may be patched
+	// if and only if both of following conditions are met, we can just return
+	// 1. canary service has been already cleaned up
+	// 2. stable service has no selector
+	stableService := &corev1.Service{}
+	err = m.Get(context.TODO(), client.ObjectKey{Namespace: c.Namespace, Name: trafficRouting.Service}, stableService)
+	stableServiceRestored := errors.IsNotFound(err) || (err == nil && stableService.Spec.Selector[c.RevisionLabelKey] == "")
 	// if canary svc has been already cleaned up, just return
 	// even DisableGenerateCanaryService is true, canary svc still exists, because canary service is stable service
-	if err = m.Get(context.TODO(), client.ObjectKeyFromObject(cService), cService); err != nil {
+	if err = m.Get(context.TODO(), client.ObjectKeyFromObject(cService), cService); err != nil && stableServiceRestored {
```

Comment:
if err != nil && !errors.IsNotFound(err), the logic will skip the if clause and start restore stable service, it seems better to just return error and retry

Reply:
it's ok because we will get the stable Service again in the following restoring stable service
whatever, i have refactored this logic, looking forward to your check again, many thanks!